tests/unit/testsuite/Bolt/Boltpay/Block/Checkout/BoltpayTest.php

```diff
@@ -255,6 +289,401 @@ public function testGetSaveOrderURL()
+            ],
+            [
+                # When both keys exist on product page, multi-step is used
+                "multiStepKey" => "multi+payOnly-key-on-product",
+                "paymentOnlyKey" => "payOnly+multi-key-on-product",
+                "routeName" => "catalog",
+                "controllerName" => "product",
+                "predictedReturnValue" => "multi+payOnly-key-on-product"
+            ],
+            [
+                # When only multi-step key exists on product page, null payment-only, multi-step is used
+                "multiStepKey" => "multi-key-on-product",
+                "paymentOnlyKey" => null,
+                "routeName" => "catalog",
+                "controllerName" => "product",
+                "predictedReturnValue" => "multi-key-on-product"
+            ],
+            [
+                # When only payment-only key exists on product page, null multi-step, payment-only is used
+                "multiStepKey" => null,
+                "paymentOnlyKey" => "payOnly-key-on-product",
+                "routeName" => "catalog",
+                "controllerName" => "product",
+                "predictedReturnValue" => "payOnly-key-on-product"
+            ]
+        ];
+    }
+
+    /**
+     * @test
+     * that an exception is thrown for all variants if neither a multi-step nor a payment-only key has been configured
+     *
+     * @covers ::getPublishableKeyForThisPage
+     * @dataProvider getPublishableKeyForThisPage_whenNoKeyIsConfiguredProvider
+     * @expectedException Bolt_Boltpay_BoltException
+     * @expectedExceptionMessage No publishable key has been configured.
+     *
+     * @param string|null $multiStepKey             the multi-step publishable key, empty string or null indicates not set
+     * @param string|null $paymentOnlyKey           the payment-only publishable key, empty string or null indicates not set
+     * @param string      $routeName                The magento path route name
+     * @param string      $controllerName           The magento path controller name
+     *
+     * @throws Exception if there is a problem in setting up test dependencies
+     */
+    public function getPublishableKeyForThisPage_whenNoKeyIsConfigured($multiStepKey, $paymentOnlyKey, $routeName, $controllerName)
+    {
+        $this->setUp_getPublishableKeyForThisPage($multiStepKey, $paymentOnlyKey, $routeName, $controllerName);
+
+        $this->helperMock->expects($this->once())->method('logException');
+        $this->helperMock->expects($this->once())->method('notifyException');
+
+        $actualMultiStepKey =  $this->app->getStore()->getConfig('payment/boltpay/publishable_key_multipage');
+        $actualPaymentOnlyKey = $this->app->getStore()->getConfig('payment/boltpay/publishable_key_onepage');
+
+        $this->currentMock->getPublishableKeyForThisPage();
+        $this->assertFalse($actualMultiStepKey || $actualPaymentOnlyKey);
```

Comment:
Bro, you made me cry!!!
Unit test could not contains logic!

Reply:
Hmmm.  There is probably no reason to cry here, especially if we accustomed to recognizing some general characteristics of `&&` and `||`.

Let's say that I have four conditions, `$cond1`, `$cond2`, `$cond3`, and `$cond4`.  Now, I want to **assert that all the conditions are false** in order for the test to pass.

Then the follow two approaches accomplish the same thing --

- Long Approach - 
```
$this->assertFalse($cond1);
$this->assertFalse($cond2);
$this->assertFalse($cond3);
$this->assertFalse($cond4);
```

- Short Approach -
```
$this->assertFalse($cond1 || $cond2 || $cond3 || $cond4);
```

And if I were interested in asserting that all the conditions were true, I could do this:
```
$this->assertTrue($cond1 && $cond2 && $cond3 && $cond4);
```

If I wanted to be verbose, I could add a message to the assertion:
```
$this->assertFalse($cond1 || $cond2 || $cond3 || $cond4, "All should be false: $cond1,$cond2,$cond3,$cond4, " );
```